File: src/dynamic-programming/hard/maximize-expression/c-sharp/brute-force.cs
using System;

/*
The brute-force approach to solving this problem is to simply iterating through every valid choice of ABCD and to evaluate the expression at each iteration, while keeping track of the maximum value. Just return the maximum after considering all possibilities. 

While this solution results in optimal space complexity, it runs in O(n^4) which is pretty bad.

Time : O(n^4) - Where N is the length of the input array
Space: O(1)   - Calculations done in place
*/
public class Program
{
    public int MaximizeExpression(int[] array)
    {
        if (array.Length < 4)
        {
            return 0;
        }

        var max = int.MinValue;

        for (var a = 0; a < array.Length; a++)
        {
            for (var b = a + 1; b < array.Length; b++)
            {
                for (var c = b + 1; c < array.Length; c++)
                {
                    for (var d = c + 1; d < array.Length; d++)
                    {
                        var current = EvaluateExpression(array[a], array[b], array[c], array[d]);
                        max = Math.Max(current, max);
                    }
                }
            }
        }
        return max;
    }

    private int EvaluateExpression(int a, int b, int c, int d)
    {
        return a - b + c - d;
    }
}

File: src/strings/extreme/longest-balanced-substring/go/brute-force.go
package main

// O(n^3) time | O(n) space - where N is the length of the input string
func LongestBalancedSubstring(str string) int {
	maxLength := 0
	for i := range str {
		for j := i + 2; j < len(str)+1; j++ {
			if isBalanced(str[i:j]) {
				currentLength := j - i
				maxLength = max(maxLength, currentLength)
			}
		}
	}
	return maxLength
}

func isBalanced(str string) bool {
	openParensStack := []rune{}
	for _, char := range str {
		if char == '(' {
			openParensStack = append(openParensStack, char)
		} else if len(openParensStack) > 0 {
			openParensStack = openParensStack[:len(openParensStack)-1]
		} else {
			return false
		}
	}
	return len(openParensStack) == 0
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
